File: chp01_Singleton/Swift/Testing_DatabaseConnection_reset.swift
import Foundation

class DatabaseConnection {
    private static var _sharedInstance: DatabaseConnection?
    
    internal static func _initializeInstance() {
        _sharedInstance = DatabaseConnection()
    }

    static var sharedInstance: DatabaseConnection {
        guard let instance = _sharedInstance else {
            _initializeInstance(); return _sharedInstance!
        }
        return instance
    }
    
    private init() {}
    
    func connectToDatabase() {
        // Connect to the database here
        print("Connected to database")
    }
    
    func executeQuery(query: String) {
        // Execute the query on the database
        print("Executing query: \(query)")
    }
    
    static func resetInstance() {
        _sharedInstance = nil
        print("DatabaseConnection instance reset for testing purposes")
    }
}

// Test class using Dependency Injection
class DatabaseConnectionTest {
    private let databaseConnection: DatabaseConnection
    
    init(databaseConnection: DatabaseConnection) {
        self.databaseConnection = databaseConnection
    }
    
    func runTest() {
        // Example test case: connect to database and execute a query
        databaseConnection.connectToDatabase()
        databaseConnection.executeQuery(query: "SELECT * FROM test_table")
    }
}

// Usage
let dbConnection = DatabaseConnection.sharedInstance
dbConnection.connectToDatabase()
dbConnection.executeQuery(query: "SELECT * FROM users")

// Dependency Injection Example
let dbTest = DatabaseConnectionTest(databaseConnection: dbConnection)
dbTest.runTest()

// Reset Singleton Instance for Testing Purposes
DatabaseConnection.resetInstance()
// The next line will crash intentionally to demonstrate reset behavior:
// let dbConnection2 = DatabaseConnection.sharedInstance
// Instead, reinitialize to prevent crash
DatabaseConnection._initializeInstance()
let dbConnection2 = DatabaseConnection.sharedInstance
dbConnection2.connectToDatabase()
dbConnection2.executeQuery(query: "SELECT * FROM new_users")

File: chp01_Singleton/Kotlin/Testing_DatabaseConnection_reset.kt
object DatabaseConnection {
    @Volatile
    private var instance: DatabaseConnection? = null

    init {
        // Initialize the database connection here
        println("Database connection initialized")
    }

    fun executeQuery(query: String): String {
        // Execute the query and return the result
        println("Executing query: $query")
        return "Query executed successfully"
    }

    @Synchronized
    fun resetInstance() {
        instance = null
        println("Database connection instance reset for testing purposes")
    }
}

// Test class using Dependency Injection
class DatabaseConnectionTest(private val databaseConnection: DatabaseConnection) {
    fun runTest() {
        // Example test case: execute a query
        val result = databaseConnection.executeQuery("SELECT * FROM test_table")
        println("Test Query Result: $result")
    }
}

fun main() {
    val dbInstance = DatabaseConnection
    val result = dbInstance.executeQuery("SELECT * FROM users")
    println(result)

    // Dependency Injection Example
    val dbTest = DatabaseConnectionTest(dbInstance)
    dbTest.runTest()

    // Reset the Singleton Instance for Testing Purposes
    DatabaseConnection.resetInstance()
    val dbInstance2 = DatabaseConnection
    val result2 = dbInstance2.executeQuery("SELECT * FROM new_users")
    println(result2)
}
